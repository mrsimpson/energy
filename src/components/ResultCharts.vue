<script setup lang="ts">
import { computed, toRefs } from "vue";
import { Bar } from "vue-chartjs";
import {
  Chart as ChartJS,
  Title,
  Tooltip,
  Legend,
  BarElement,
  CategoryScale,
  LinearScale,
type TooltipItem,
} from "chart.js";
import { euros } from "@/lib/Numbers";

ChartJS.register(Title, Tooltip, Legend, BarElement, CategoryScale, LinearScale);

type Year = number;
type Amount = number;

const props = defineProps<{
  years: Year[];
  bills: Amount[];
  subsidization: Amount[];
  savings: Amount[];
}>();

const { bills, subsidization, savings } = toRefs(props);

// make the chart reactive on all props
const series = [
  {
    name: "Eigenanteil",
    data: bills,
    color: "rgba(255, 99, 132, 0.2)",
    borderColor: "#dbb",
  },
  {
    name: "Staatliche Unterst√ºtzung",
    data: subsidization,
    color: "rgba(255, 205, 86, 0.2)",
    borderColor: "rgba(255, 205, 86)",
  },
  {
    name: "Eigene Einsparung",
    data: savings,
    color: "rgba(201, 203, 207, 0.05)",
    borderColor: "rgb(201, 203, 207)",
  },
];

const scales = {
  x: {
    stacked: true,
  },
  y: {
    stacked: true,
    ticks: {
      callback(value: string | number) {
        return euros(value as number, 0)
      }
    }
  },
};

const options = {
  scales,
  borderSkipped: "middle",
  plugins: {
    tooltip: {
      callbacks: {
        label(context: TooltipItem<"bar">) {
          return context.dataset.label + ": " + euros(context.parsed.y);
        },
      },
    },
  }
};

const chartData = computed(() => ({
  labels: [2021, 2022, 2023],
  datasets: series.map((s) => ({
    data: s.data.value,
    label: s.name,
    backgroundColor: s.color,
    borderColor: s.borderColor,
    borderWidth: 2,
  })),
}));
</script>

<template>
  <Bar
    :chart-data="chartData"
    :chart-options="options"
    :height="200"
  />
</template>

<style scoped>
.result-chart canvas {
  height: 400px
}
</style>
