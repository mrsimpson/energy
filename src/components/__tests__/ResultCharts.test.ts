import { shallowMount } from '@vue/test-utils';
import { describe, expect, it } from 'vitest';
import ResultCharts from '../ResultCharts.vue';

describe("ResultCharts", () => {
  it("should render as expected", () => {
    const wrapper = shallowMount(ResultCharts, {
      propsData: {
        years: ['2021', '2022', '2023 ohne Einsparung', '2023'],
        bills: [1000, 2000, 3000, 2500],
        subsidization: [0, 200, 500, 500],
        savings: [0, 0, 0, 500]
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
