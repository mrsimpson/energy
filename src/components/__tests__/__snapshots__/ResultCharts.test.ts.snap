// Vitest Snapshot v1

exports[`ResultCharts > should render as expected 1`] = `
<vue-apex-chart-stub
  height="auto"
  options="[object Object]"
  series="[object Object],[object Object],[object Object]"
  type="bar"
  width="500"
/>
`;
